File: DiaryNutrition/app/src/main/java/ediger/diarynutrition/weight/WeightViewModel.java
package ediger.diarynutrition.weight;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Transformations;

import java.util.Calendar;
import java.util.List;

import ediger.diarynutrition.AppContext;
import ediger.diarynutrition.PreferenceHelper;
import ediger.diarynutrition.data.DiaryRepository;
import ediger.diarynutrition.data.source.entities.Weight;
import ediger.diarynutrition.objects.SingleLiveEvent;

public class WeightViewModel extends AndroidViewModel {

    private DiaryRepository mRepository;

    private LiveData<List<Weight>> mWeight;

    private MutableLiveData<Calendar> mDateSince = new MutableLiveData<>();

    private LiveData<List<Weight>> mWeightSinceDate;

    public WeightViewModel(@NonNull Application application) {
        super(application);
        mRepository = ((AppContext) application).getRepository();

        mWeight = mRepository.getWeight();

        Calendar from = Calendar.getInstance();
        from.add(Calendar.DAY_OF_YEAR, -8);
        mDateSince.setValue(from);

        mWeightSinceDate = Transformations
                .switchMap(mDateSince, date -> mRepository.getWeight(date));
    }

    public LiveData<List<Weight>> getWeight() {
        return mWeight;
    }

    public Weight getRecentWeight() {
        return mRepository.getRecentWeight();
    }

    LiveData<List<Weight>> getWeightSinceDate() {
        return mWeightSinceDate;
    }

    public void addWeight(Weight weight) {
        Calendar day = Calendar.getInstance();
        day.setTimeInMillis(weight.getDatetime());

        Weight lastWeight = mRepository.getWeightForDay(day);

        if (lastWeight == null) {
            mRepository.addWeight(weight);
        } else {
            lastWeight.setAmount(weight.getAmount());
            mRepository.updateWeight(lastWeight);
        }

        //Updates for use of NutritionProgram
        PreferenceHelper.setValue(PreferenceHelper.KEY_WEIGHT, weight.getAmount());
    }

    void deleteWeight(int id) {
        mRepository.deleteWeight(id);
    }

    void setDate(Calendar since) {
        if (since != null) {
            mDateSince.setValue(since);
        }
    }

}

File: DiaryNutrition/app/src/main/java/ediger/diarynutrition/weight/WeightViewModel.kt
package ediger.diarynutrition.weight

import android.app.Application
import androidx.lifecycle.*
import ediger.diarynutrition.AppContext
import ediger.diarynutrition.KEY_WEIGHT
import ediger.diarynutrition.PreferenceHelper
import ediger.diarynutrition.data.source.entities.Weight
import kotlinx.coroutines.launch
import java.util.*

class WeightViewModel(application: Application) : AndroidViewModel(application) {

    private val weightRepository = (application as AppContext).weightRepository

    val weight: LiveData<List<Weight>> = weightRepository.getWeight()

    private val mDateSince = MutableLiveData<Calendar>()

    val weightSinceDate: LiveData<List<Weight>>

    init {
        val from = Calendar.getInstance().apply { add(Calendar.DAY_OF_YEAR, -8) }
        mDateSince.value = from
        weightSinceDate = Transformations
                .switchMap(mDateSince) { date: Calendar -> weightRepository.getWeight(date) }
    }

    fun addWeight(weight: Weight) = viewModelScope.launch {
        val day = Calendar.getInstance().apply { timeInMillis = weight.datetime }
        val lastWeight = weightRepository.getWeightForDay(day)

        if (lastWeight == null) {
            weightRepository.addWeight(weight)
        } else {
            lastWeight.amount = weight.amount
            weightRepository.updateWeight(lastWeight)
        }

        //Updates for use of NutritionProgram
        PreferenceHelper.setValue(KEY_WEIGHT, weight.amount)
    }

    fun deleteWeight(id: Int) = viewModelScope.launch { weightRepository.deleteWeight(id) }

    fun setDate(since: Calendar) {
        mDateSince.value = since
    }
}